### DISCUSSION TITLE
cwd property for spawned child processes

### DISCUSSION THREAD

**Discussion by prettydiff** [nodejs/help#5016] [Q&A]
### What is the problem this feature will solve?

The current working directory of a Node parent process is conveniently available as `process.cwd()`, but there is no such solution for a child process. If a command changes the directory of a child process there is no messaging from child process to parent indicating that change and there is no Node solution to ask what the child's current directory is.

### What is the feature you are proposing to solve the problem?

I request a `cwd()` method for spawned child processes.  Example:

```typescript
import { ChildProcess, spawn } from "child_process";

const handler = function(output:Buffer):void {
        console.log(output.toString());
        console.log(child.cwd());
    },
    instruction:string = "node other_program",
    child:ChildProcess = spawn(instruction, [], {
        cwd: process.cwd(),
        shell: true
    });
child.stdin.on("data", handler);
```

### What alternatives have you considered?

I have looked into writing my own solution by writing to the child process's stdin, but this is wildly complicated cross OS and it is asynchronous which results in timing variations apart from stdin/stdout messaging.

**Comment by cjihrig** [comment]
If you use `fork()` you can send messages back and forth between the parent and child using IPC.

Adding a `cwd()` for arbitrary child processes is probably not going to be something that gets implemented though. I don't think it can be done cross platform (but I'm happy to be wrong).

**Comment by prettydiff** [comment]
@cjihrig What message should I pass to the forked process in order to discover the cwd?

**Comment by cjihrig** [comment]
That would be up to you. You'll fork a Node process with `fork()`. Then, you could use [`subprocess.send()`](https://nodejs.org/api/child_process.html#subprocesssendmessage-sendhandle-options-callback) to send messages to the child process, or [`process.send()`](https://nodejs.org/api/process.html#processsendmessage-sendhandle-options-callback) to send messages to the parent process. You could have the parent query the child process and have it respond, or just have the child send messages without being queried, but that's where you would include the child's pwd.

**Comment by prettydiff** [comment]
What if this feature request were limited to child processes with a *shell* option set to true?  A shell will know what its current directory is.  That way the same code that is used for the `process.cwd()` of the parent can be reused for the child, which I believe is: https://github.com/nodejs/node/blob/main/src/node_process_methods.cc#L127-L141

**Comment by cjihrig** [comment]
It won't work with the `shell` option or anything that is not a Node.js process. As far as I know, this is not something we can do cross platform. Have you had any luck programmatically getting a child process cwd like this in a runtime other than Node (Python, Go, etc.)?

**Comment by prettydiff** [comment]
No, I have not had any luck.  I am writing a command terminal interface in a browser OS I am writing.  I can already keep track of cwd for a user performing file system traversal, because everything executes through the current parent process where I am aware of state of all artifacts.  Child processes are tricky because I have no idea what a user executed script might do, so when I report output to the user I have to find some way of determining what a given shell's current directory is only after the child process writes to stdout but before the next round of stdout

**Comment by bnoordhuis** [comment]
I'll convert this to a discussion because, like Colin said, this can't be implemented in a cross-platform manner.

Even if it could, it'd be pretty useless because of how race-y it'd be: by the time you know the child's working directory, it may have chdir'd again.

**Comment by KidusB9** [comment]
It seems that you are proposing the addition of a cwd() method to the ChildProcess class in the child_process module of Node.js. This method would allow you to retrieve the current working directory of a child process, similar to the process.cwd() method that is available for the parent process.

One potential alternative to this solution would be to communicate the current working directory of the child process to the parent process through some other means, such as through a message sent over stdout or through a separate communication channel. However, this approach would require more effort and coordination between the parent and child processes, and may not be as convenient as having a dedicated method for retrieving the current working directory of the child process.
